core-to-plc/src/Language/Plutus/CoreToPLC/Compiler/Builtins.hs

```diff
@@ -73,19 +73,42 @@ Haskell. But there is a problem: they use either primitive or Scott-encoded vers
 whereas when we compile them from Haskell they will end up as abstract types, and so the types
 won't line up at the call site.
 
+That is, we generate something like this:
+(/\ (Int :: *) .
+  (\ addInteger : Int -> Int -> Int .
+      <use addInteger>
+  )
+  (\ x,y : Int . <builtin addInteger> x y) -- Uh oh, type error, can't show that Int = <builtin int>!
+)
+{<builtin int>}
+
 We handle this in two different ways:
-- For the types like Bool and Unit which are really algebraic type, and which have constructors etc.
+- For the types like Bool and Unit which are really algebraic types, and which have constructors etc.
 that we care about elsewhere, we insert adaptor code into the definition of the builtin (see note [Mapping builtins]).
 - For types like Int and Bytestring which don't have visible constructors, we can treat them as completely opaque,
 and we use a type alias. (Actually we fake the alias until we have aliases in PIR)
 
-Why do we need an alias rather than just a normal type def? The problem is that the type we want to present to the
-user for `addInteger` is `Int -> Int -> Int`, but in the *body* of the definition the argument types need to be
-*PLC* ints otherwise the builtin won't be happy. So we need to break the abstraction in the body, hence the alias.
+Here's how that looks for a primitive that takes Ints and returns a Boolean, assuming we have bound Int and Bool
+already as an alias and an abstract type respectively:
+(\ equalsInteger : Int -> Int -> Bool .
+  <use equalsInteger>
+)
+(\ x, y : Int . -- No need to do anything to the arguments, since we're using a transparent alias for Int
```

Comment:
Great examples, thank you. I now understand most of what is written, but this part confuses me a bit. I would expect it to be

```
(/\ (Int :: *) .
    (\ equalsInteger : Int -> Int -> Bool .
      <use equalsInteger>
    )
{<builtin int>}
(\ x, y : Int .
    (<builtin equalsInteger> x y) {Bool} True False
)
```

like the one below. There is a part saying `assuming we have bound Int`, but I don't think you can have `Int` bound and supply `equalsInteger` like that? I mean, it's like having 

```
(/\ (Int :: *) .
  (\ addInteger : Int -> Int -> Int . -- Type signature is fine inside the abstraction
      <use addInteger>
  )
  (\ x,y : <builtin int> . <builtin addInteger> x y) -- Type error
)
{<builtin int>}
```

instead of what is written below. Or am I just misunderstanding something?

Reply:
The crucial bit is what I said before - we bind `Int` *as a transparent type alias*. I'm handwaving a bit here, because what that means in practice is that we inline the definition everywhere, but if we got type alias support in PIR then we could add it as a type alias.

Reply:
But why do we do so for `Int`? Is it not possible to keep `Int` always abstract? I mean, I do not understand at the moment why the fact that we return a `Bool` has impact on how we pass `Int`s in.

Reply:
Because otherwise we would have to do it in this interleaved way, and PIR doesn't support that except specifically for things defined as datatypes.